### PR TITLE
feat: support reloading workflows at runtime

### PR DESCRIPTION
**Pull request by didier-wenzek**
## Proposed changes

In order to support dynamic reloading of workflow, without breaking a running workflow,
the proposal is to:

- use a hash of the workflow definition file to distinguish workflow versions
- persist a copy of each version for a given operation
- make the copy when a command instance is triggered
- use reference-counting to remove copies when no more in-use.

Plan:

- [x] Attach a version to each workflow, this version being a hash of the definition file.
- [x] Attach a version to each operation instance, taking the version of the workflow definition in use.
- [x] Refactor the agent to group the workflow loading logic in a single place.
- [x] When a new operation instance is created, copy its definition in a persisted directory of workflows in-use. 
- [x] On-start load not only the user-provided workflow definition, but also those of the workflows in-use.
- [x] Reload workflow definitions on file change using inotify. 
- [x] On file change, reload the workflow definition (without erasing any definition still in-use).
- [x] When a user-defined workflow is removed, clear the associated capability (letting any still running operation run to completion).
- [x] When a builtin workflow specialization is removed, restore the builtin version if any.
- [x] When a workflow definition is removed, keep a copy of that definition if still in-use.  

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/3156

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_agent/src/operation_workflows/persist.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        }
+    }
+}
+
+async fn read_operation_workflow(
+    path: &Path,
+) -> Result<(OperationWorkflow, WorkflowVersion), anyhow::Error> {
+    let bytes = tokio::fs::read(path).await.context("Fail to read file")?;
+    let input = std::str::from_utf8(&bytes).context("Fail to extract UTF8 content")?;
+    let version = sha256::digest(input);
```
Comment (didier-wenzek) [inline crates/core/tedge_agent/src/operation_workflows/persist.rs:101]: Using `md5` should be enough here - as there is no crypto concerns.

**Review by gligorisaev — COMMENTED** [review]
The Test is testing the described feature
# 📊 **Consolidated Test Summary for 10 Runs**


### 🛠️ **Test Results Summary - Run 1**

**Test Suite**: Dynamic Workflow Reloading  
**Date**: 2024-10-17 14:33:18  
**Total Tests**: 7  
**Passed**: 7  
**Failed**: 0  
**Duration**: 37.56 seconds  
**Success Rate**: 100.00%

#### 📄 **Test Details**:
- Test: Create User-Defined Operation - Status: PASSED
- Test: Update User-Defined Operation - Status: PASSED
- Test: Remove User-Defined Operation - Status: PASSED
- Test: Updating A Workflow Twice Before Using It - Status: PASSED
- Test: Override Builtin Operation - Status: PASSED
- Test: Recover Builtin Operation - Status: PASSED
- Test: Trigger Workflow Update From A Main Workflow - Status: PASSED

### 🛠️ **Test Results Summary - Run 2**

**Test Suite**: Dynamic Workflow Reloading  
**Date**: 2024-10-17 14:33:55  
**Total Tests**: 7  
**Passed**: 7  
**Failed**: 0  
**Duration**: 34.62 seconds  
**Success Rate**: 100.00%

#### 📄 **Test Details**:
- Test: Create User-Defined Operation - Status: PASSED
- Test: Update User-Defined Operation - Status: PASSED
- Test: Remove User-Defined Operation - Status: PASSED
- Test: Updating A Workflow Twice Before Using It - Status: PASSED
- Test: Override Builtin Operation - Status: PASSED
- Test: Recover Builtin Operation - Status: PASSED
- Test: Trigger Workflow Update From A Main Workflow - Status: PASSED

### 🛠️ **Test Results Summary - Run 3**

**Test Suite**: Dynamic Workflow Reloading  
**Date**: 2024-10-17 14:34:31  
**Total Tests**: 7  
**Passed**: 7  
**Failed**: 0  
**Duration**: 34.70 seconds  
**Success Rate**: 100.00%

#### 📄 **Test Details**:
- Test: Create User-Defined Operation - Status: PASSED
- Test: Update User-Defined Operation - Status: PASSED
- Test: Remove User-Defined Operation - Status: PASSED
- Test: Updating A Workflow Twice Before Using It - Status: PASSED
- Test: Override Builtin Operation - Status: PASSED
- Test: Recover Builtin Operation - Status: PASSED
- Test: Trigger Workflow Update From A Main Workflow - Status: PASSED

### 🛠️ **Test Results Summary - Run 4**

**Test Suite**: Dynamic Workflow Reloading  
**Date**: 2024-10-17 14:35:08  
**Total Tests**: 7  
**Passed**: 7  
**Failed**: 0  
**Duration**: 35.52 seconds  
**Success Rate**: 100.00%

#### 📄 **Test Details**:
- Test: Create User-Defined Operation - Status: PASSED
- Test: Update User-Defined Operation - Status: PASSED
- Test: Remove User-Defined Operation - Status: PASSED
- Test: Updating A Workflow Twice Before Using It - Status: PASSED
- Test: Override Builtin Operation - Status: PASSED
- Test: Recover Builtin Operation - Status: PASSED
- Test: Trigger Workflow Update From A Main Workflow - Status: PASSED

### 🛠️ **Test Results Summary - Run 5**

**Test Suite**: Dynamic Workflow Reloading  
**Date**: 2024-10-17 14:35:43  
**Total Tests**: 7  
**Passed**: 7  
**Failed**: 0  
**Duration**: 32.91 seconds  
**Success Rate**: 100.00%

#### 📄 **Test Details**:
- Test: Create User-Defined Operation - Status: PASSED
- Test: Update User-Defined Operation - Status: PASSED
- Test: Remove User-Defined Operation - Status: PASSED
- Test: Updating A Workflow Twice Before Using It - Status: PASSED
- Test: Override Builtin Operation - Status: PASSED
- Test: Recover Builtin Operation - Status: PASSED
- Test: Trigger Workflow Update From A Main Workflow - Status: PASSED

### 🛠️ **Test Results Summary - Run 6**

**Test Suite**: Dynamic Workflow Reloading  
**Date**: 2024-10-17 14:36:19  
**Total Tests**: 7  
**Passed**: 7  
**Failed**: 0  
**Duration**: 34.90 seconds  
**Success Rate**: 100.00%

#### 📄 **Test Details**:
- Test: Create User-Defined Operation - Status: PASSED
- Test: Update User-Defined Operation - Status: PASSED
- Test: Remove User-Defined Operation - Status: PASSED
- Test: Updating A Workflow Twice Before Using It - Status: PASSED
- Test: Override Builtin Operation - Status: PASSED
- Test: Recover Builtin Operation - Status: PASSED
- Test: Trigger Workflow Update From A Main Workflow - Status: PASSED

### 🛠️ **Test Results Summary - Run 7**

**Test Suite**: Dynamic Workflow Reloading  
**Date**: 2024-10-17 14:36:55  
**Total Tests**: 7  
**Passed**: 7  
**Failed**: 0  
**Duration**: 34.91 seconds  
**Success Rate**: 100.00%

#### 📄 **Test Details**:
- Test: Create User-Defined Operation - Status: PASSED
- Test: Update User-Defined Operation - Status: PASSED
- Test: Remove User-Defined Operation - Status: PASSED
- Test: Updating A Workflow Twice Before Using It - Status: PASSED
- Test: Override Builtin Operation - Status: PASSED
- Test: Recover Builtin Operation - Status: PASSED
- Test: Trigger Workflow Update From A Main Workflow - Status: PASSED

### 🛠️ **Test Results Summary - Run 8**

**Test Suite**: Dynamic Workflow Reloading  
**Date**: 2024-10-17 14:37:32  
**Total Tests**: 7  
**Passed**: 7  
**Failed**: 0  
**Duration**: 34.53 seconds  
**Success Rate**: 100.00%

#### 📄 **Test Details**:
- Test: Create User-Defined Operation - Status: PASSED
- Test: Update User-Defined Operation - Status: PASSED
- Test: Remove User-Defined Operation - Status: PASSED
- Test: Updating A Workflow Twice Before Using It - Status: PASSED
- Test: Override Builtin Operation - Status: PASSED
- Test: Recover Builtin Operation - Status: PASSED
- Test: Trigger Workflow Update From A Main Workflow - Status: PASSED

### 🛠️ **Test Results Summary - Run 9**

**Test Suite**: Dynamic Workflow Reloading  
**Date**: 2024-10-17 14:38:10  
**Total Tests**: 7  
**Passed**: 7  
**Failed**: 0  
**Duration**: 36.98 seconds  
**Success Rate**: 100.00%

#### 📄 **Test Details**:
- Test: Create User-Defined Operation - Status: PASSED
- Test: Update User-Defined Operation - Status: PASSED
- Test: Remove User-Defined Operation - Status: PASSED
- Test: Updating A Workflow Twice Before Using It - Status: PASSED
- Test: Override Builtin Operation - Status: PASSED
- Test: Recover Builtin Operation - Status: PASSED
- Test: Trigger Workflow Update From A Main Workflow - Status: PASSED

### 🛠️ **Test Results Summary - Run 10**

**Test Suite**: Dynamic Workflow Reloading  
**Date**: 2024-10-17 14:38:49  
**Total Tests**: 7  
**Passed**: 7  
**Failed**: 0  
**Duration**: 37.12 seconds  
**Success Rate**: 100.00%

#### 📄 **Test Details**:
- Test: Create User-Defined Operation - Status: PASSED
- Test: Update User-Defined Operation - Status: PASSED
- Test: Remove User-Defined Operation - Status: PASSED
- Test: Updating A Workflow Twice Before Using It - Status: PASSED
- Test: Override Builtin Operation - Status: PASSED
- Test: Recover Builtin Operation - Status: PASSED
- Test: Trigger Workflow Update From A Main Workflow - Status: PASSED

File: crates/core/tedge_api/src/workflow/supervisor.rs [hunk 2]
```diff
@@ -229,6 +298,152 @@ impl WorkflowSupervisor {
...
+/// - The current version, i.e. the version to be used for new operation instances.
+/// - One in-use version, that was used by some operation instance when the current version has been updated.
+/// - The builtin version provided by the agent and which can be overridden by the users.
+///
+/// None of these versions are mandatory. For instance, an operation can have no builtin version,
+/// no current version (because the user removed the definition file), but a version still in-use
+/// (that has been started before the user deprecated the operation).
+struct WorkflowVersions {
+    operation: OperationName,
+    current: Option<WorkflowVersion>,
```
Comment (albinsuresh) [inline crates/core/tedge_api/src/workflow/supervisor.rs:314]: While reviewing the code, I was often getting `current` and `in-use` confused and mixed up, esp with `current` as it could be wrongly interpreted as the "current one being used". Calling it something like `latest` might help.

File: crates/core/tedge_agent/src/operation_workflows/persist.rs [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+                .register_builtin_workflow(capability.as_str().into())
+            {
+                error!("Fail to register built-in workflow for {capability} operation: {err}");
+            }
+        }
+    }
+
+    /// Update the workflow definitions after some on-disk changes
+    ///
+    /// Return the operation name when the operation has been deprecated.
```
Comment (albinsuresh) [inline crates/core/tedge_agent/src/operation_workflows/persist.rs:162]:
```suggestion
    /// Return the operation capability deregistration message when the operation has been deprecated.
```

**Comment by reubenmiller** [comment]
> The Test is testing the described feature
> 
> Tested on RPI 4:
> 
> # 📊 **Consolidated Test Summary for 10 Runs**


@gligorisaev it would be more efficient to use the flakefinder task, and just post the summary from the console output (just the last 3 lines is fine). Posting the results of all the tests is unnecessary (duplicate information) and just clouds the ticket. In the future we could even look at creating a CI workflow which can do on-demand flake finder runs on specific suites or tagged tests.

By using flakefinder, it is easier for other people to run to check that any newly added suite is (hopefully) flake-free. If there is a detected "flake" then the developer will also need to know how to re-run the same tests to help them with debugging.

In this case, running the following command did the same thing as you did (though this was run inside a container):

```sh
invoke flake-finder --suite "dynamic_workflow_reloading" --iterations 10 --outputdir output_ff --clean
```

**Output (only the last 3 lines)**

```sh
Overall: PASSED
Results: 10 iterations, 10 passed, 0 failed
Elapsed time: 0:05:46.354385
```

**Review by albinsuresh — COMMENTED** [review]
LGTM. Just a few minor suggestions on possible simplifications (if feasible).

File: crates/core/tedge_agent/src/state_repository/state.rs [hunk 4]
```diff
@@ -17,30 +17,41 @@ pub struct AgentStateRepository<T> {
...
+    // Check that the given directory is actually writable, by creating an empty test file
+    let test_file = state_dir.join(state_dir.join(".--test--"));
+    match File::create(test_file.clone()).and_then(|mut file| file.write_all(b"")) {
+        Ok(_) => {
+            let _ = std::fs::remove_file(test_file);
+            state_dir
+        }
+        Err(err) => {
+            warn!("Cannot use {state_dir:?} to store tedge-agent state: {err}");
+            agent_default_state_dir(tedge_root)
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/state_repository/state.rs:39]: This directory is critical for A/B firmware update.

See the rational here: https://github.com/thin-edge/thin-edge.io/issues/2488

File: crates/core/tedge_api/src/workflow/supervisor.rs (hunk 2 above)
Comment (didier-wenzek) [reply crates/core/tedge_api/src/workflow/supervisor.rs:314]: Unfortunately, `latest` is used for the on-disk version.

- `current` is the version currently loaded by the workflow engine.
- `in-use` is a copy that is kept for a running workflow in case the agent restarts while the operation is still under execution. On restart that version must be used and not a new one.

File: crates/core/tedge_agent/src/operation_workflows/persist.rs [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+    state_dir: Utf8PathBuf,
+
+    // Map each workflow version to its workflow file
+    //
+    // For a fresh new workflow definition, this points to the user-defined file
+    // When the workflow definition is in use, this points to a copy in the state directory.
+    definitions: HashMap<WorkflowVersion, (OperationName, Utf8PathBuf)>,
+    in_use_copies: HashMap<OperationName, WorkflowVersion>,
+
+    // The in-memory representation of all the workflows (builtin, user-defined, i,n-use).
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/operation_workflows/persist.rs]: https://github.com/thin-edge/thin-edge.io/pull/3180/commits/f0e154c13fcc71a4b63c7c48cb7cca15c13481e8

File: crates/core/tedge_agent/src/operation_workflows/persist.rs (hunk 3 above)
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/operation_workflows/persist.rs]: https://github.com/thin-edge/thin-edge.io/pull/3180/commits/3ecdaa6bc4fe405ae93ef4871eeda5a1a17dc23e

File: crates/core/tedge_api/src/workflow/supervisor.rs [hunk 6]
```diff
@@ -229,6 +298,152 @@ impl WorkflowSupervisor {
...
+    builtin: Option<WorkflowVersion>,
+    in_use: Option<WorkflowVersion>,
+    versions: HashMap<WorkflowVersion, OperationWorkflow>,
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/core/tedge_api/src/workflow/supervisor.rs]: Done: https://github.com/thin-edge/thin-edge.io/pull/3180/commits/e512dd77a77bd45566ca0b755c6abebb0b559e8a

Thank @albinsuresh, this was a really neat suggestion.

File: crates/core/tedge_api/src/workflow/supervisor.rs [hunk 7]
```diff
@@ -229,6 +298,116 @@ impl WorkflowSupervisor {
     }
 }
 
+/// A stack of known versions for a workflow
+///
+/// In practice, one might have 3 concurrent versions:
+///
+/// - The current version, i.e. the version to be used for new operation instances.
+/// - One in-use version, that was used by some operation instance when the current version has been updated.
+/// - The builtin version provided by the agent and which can be overridden by the users.
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/core/tedge_api/src/workflow/supervisor.rs:307]: Since we don't allow multiple operations of the same type to run concurrently, I don't think this can happen.

**Review by albinsuresh — APPROVED** [review]
LGTM.

I still have a feeling that the `@version` field and the associated lookup logic could still be simplified. As far as I understood, the SHA256 version value doesn't really have any relevance. The value `builtin` implies built-in operation. The absence of value implies the latest version on disk. So, any other value implies a cached `in-use` copy at `/etc/tedge/.agent`. So, I'm wondering why not just use some static value like `in-use` or `cached` to represent the same? That simplifies the lookup into the `WorkflowVersions` struct as well. WDYT?

File: crates/core/tedge_api/src/workflow/supervisor.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/core/tedge_api/src/workflow/supervisor.rs:307]: > we don't allow multiple operations of the same type to run concurrently

Yes, this is the assumption on which this work is based upon.

But, I'm afraid that @Bravo555 raised a valid concern.

- For the builtin operations, the commands are processed one after the other and the assumption is correct.
- But for user-defined operations, there is no mechanism to block a new command till the previous is done.
- A command instance topic cannot be re-used, but the two instances can run concurrently for the same command.
- => I will add a system test for @Bravo555 scenario.
- => I will have to fix the workflow engine to prevent concurrent runs for a given command.

**Comment by didier-wenzek** [comment]
There are two ways to solve the issue of concurrent in-use versions for an operation workflow:
1. Enforce that the operation instances for a given operation run sequentially
2. Revisit this PR to cope with several versions concurrently in use for the same operation.

We will go with the first option.

Indeed, in a near future, the plan is to give the users a way to control the concurrency of running operations.
The idea is to have too kind of operations (*update* and *request*);
so at a given time at most one *update* operation could be running
and with no limit on the number of *request* operations.

Hence, the simpler is to revisit this PR to cope with several versions concurrently in use for the same operation
Concurrency control will be then added on top of that to prevent concurrent update operations.

**Comment by reubenmiller** [comment]
> There are two ways to solve the issue of concurrent in-use versions for an operation workflow:
> 
> 1. Enforce that the operation instances for a given operation run sequentially
> 2. Revisit this PR to cope with several versions concurrently in use for the same operation.
> 
> We will go with the first option.
> 
> Indeed, in a near future, the plan is to give the users a way to control the concurrency of running operations. The idea is to have too kind of operations (_update_ and _request_); so at a given time at most one _update_ operation could be running and with no limit on the number of _request_ operations.
> 
> Hence, the simpler is to revisit this PR to cope with several versions concurrently in use for the same operation Concurrency control will be then added on top of that to prevent concurrent update operations.

Here is the link to the roadmap item to control the execution of the operations (e.g. either serial or concurrent): https://github.com/thin-edge/thin-edge.io-roadmap/issues/69

File: crates/core/tedge_api/src/workflow/supervisor.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/core/tedge_api/src/workflow/supervisor.rs]: Finally, I removed this commit.

To maintain multiple concurrent versions, yet another design was required.

See https://github.com/thin-edge/thin-edge.io/pull/3180/commits/d0cd0ebdfcdc46f960f81fc64b3da629ffb272b6

**Comment by didier-wenzek** [comment]
Here is my proposal to handle concurrent commands running the same operation but with different workflow versions.

https://github.com/thin-edge/thin-edge.io/pull/3180/commits/d0cd0ebdfcdc46f960f81fc64b3da629ffb272b6

I still have to:
- remove cached versions when no more in use
- improve the tests: there is still a race condition there

File: crates/core/tedge_agent/src/operation_workflows/persist.rs [hunk 8]
```diff
@@ -264,18 +263,15 @@ impl WorkflowRepository {
...
-        if let Some(in_use_version) = self.in_use_copies.get(operation) {
-            if in_use_version == version {
-                return;
-            }
+        if self.in_use_copies.contains(version) {
+            return;
         };
         if let Some((_, source)) = self.definitions.get(operation) {
-            let target = self.state_dir.join(operation).with_extension("toml");
+            let target = self.state_dir.join(version).with_extension("toml");
```
Comment (albinsuresh) [inline crates/core/tedge_agent/src/operation_workflows/persist.rs:270]: How about using a combination of `operation` and `version` in the file name so that we're not looking at a random hash number during debugging?

File: tests/RobotFramework/tests/tedge_agent/workflows/long-running-command-v1.toml [hunk 9]
```diff
@@ -7,7 +7,7 @@ on_success = "executing"
 [executing]
 script = "sleep ${.payload.duration}"
 on_success = "logging"
-on_error = "logging"
+on_kill = "executing"
```
Comment (albinsuresh) [inline tests/RobotFramework/tests/tedge_agent/workflows/long-running-command-v1.toml:10]: The intent of looping back to `executing` state is to resume the operation from here after the agent restart?

File: crates/core/tedge_agent/src/operation_workflows/persist.rs (hunk 8 above)
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/operation_workflows/persist.rs:270]: A `grep` should make it, but why not.

File: tests/RobotFramework/tests/tedge_agent/workflows/long-running-command-v1.toml (hunk 9 above)
Comment (didier-wenzek) [reply tests/RobotFramework/tests/tedge_agent/workflows/long-running-command-v1.toml:10]: The intent is to be sure the command is not finalized *before* the agent stops.

Currently, the actors are stopped in a random order, and I observed a run where the script was killed before the workflow engine, the latter being fast enough to move the operation forward before being stopped.

File: crates/core/tedge_api/src/workflow/supervisor.rs [hunk 10]
```diff
@@ -185,13 +186,16 @@ impl WorkflowSupervisor {
             });
         };
 
-        let version = command_state.workflow_version();
+        let Some(version) = &command_state.workflow_version() else {
+            return Err(WorkflowExecutionError::MissingVersion);
+        };
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/workflow/supervisor.rs:191]: I have to revert this change. Indeed, this might be a source of issue if the agent is updated whilst there is a pending operation. If the previous agent was not featuring command versions, then the resumed operation should not be rejected by the new agent: the current version of the workflow must be used.

File: crates/core/tedge_agent/src/operation_workflows/persist.rs (hunk 8 above)
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/operation_workflows/persist.rs]: Done along https://github.com/thin-edge/thin-edge.io/pull/3180/commits/ae56d97652643352a7e94e2b9ee73b93da3e6df6

File: crates/core/tedge_agent/src/operation_workflows/persist.rs [hunk 11]
```diff
@@ -290,6 +297,22 @@ impl WorkflowRepository {
         }
     }
 
+    async fn release_in_use_copy(&mut self, operation: &OperationName, version: &WorkflowVersion) {
+        if let Some(count) = self.in_use_copies.get_mut(version) {
+            *count -= 1;
+            if *count == 0 {
+                return;
+            }
```
Comment (didier-wenzek) [inline crates/core/tedge_agent/src/operation_workflows/persist.rs]: Oops! Fixed by https://github.com/thin-edge/thin-edge.io/pull/3180/commits/431fa64457901d5242e3729303ba4fefc979caf1

File: crates/core/tedge_agent/src/operation_workflows/persist.rs [hunk 12]
```diff
@@ -74,7 +71,7 @@ impl WorkflowRepository {
         // First, all the user-defined workflows are loaded
         let dir_path = &self.custom_workflows_dir.clone();
         if let Err(err) = self
-            .load_operation_workflows(WorkflowSource::UserDefined, dir_path)
+            .load_operation_workflows(WorkflowSource::UserDefined(dir_path))
```
Comment (albinsuresh) [inline crates/core/tedge_agent/src/operation_workflows/persist.rs:74]: Opinion: Although I understand that `WorkflowSource` being a generic type allows such generic usages, but this one with the directory path seems slightly overloaded. I agree that we can still defend it since the directory roots are also linked to the source type.

File: crates/core/tedge_agent/src/operation_workflows/persist.rs [hunk 13]
```diff
@@ -289,6 +297,22 @@ impl WorkflowRepository {
         }
     }
 
+    async fn release_in_use_copy(&mut self, operation: &OperationName, version: &WorkflowVersion) {
+        if let Some(count) = self.in_use_copies.get_mut(version) {
+            *count -= 1;
+            if *count == 0 {
```
Comment (albinsuresh) [inline crates/core/tedge_agent/src/operation_workflows/persist.rs]:
```suggestion
            if *count != 0 {
```
We should proceed with the removal only when the count become `0`, right?

File: tests/RobotFramework/tests/tedge_agent/workflows/long-running-command-v1.toml [hunk 14]
```diff
@@ -2,12 +2,16 @@ operation = "long-running-command"
 
 [init]
 action = "proceed"
-on_success = "executing"
+on_success = "scheduled"
+
+[scheduled]
+operation = "sleep"
```
Comment (albinsuresh) [inline tests/RobotFramework/tests/tedge_agent/workflows/long-running-command-v1.toml:8]:
```suggestion
operation = "sleep"     # Using a sub-operation for sleep to allow concurrent execution of multiple versions of this workflow. Using sleep command directly would have blocked each workflow, resulting in serial execution.
```

**Review by albinsuresh — APPROVED** [review]
LGTM.

File: tests/RobotFramework/tests/tedge_agent/workflows/long-running-command-v1.toml (hunk 14 above)
Comment (didier-wenzek) [reply tests/RobotFramework/tests/tedge_agent/workflows/long-running-command-v1.toml:10]: fixed by https://github.com/thin-edge/thin-edge.io/pull/3180/commits/c546f423b4220fcebcd8aad11b99c3658d5c584e

File: tests/RobotFramework/tests/tedge_agent/workflows/dynamic_workflow_reloading.robot [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+    Should Have MQTT Messages
+    ...    te/device/main///cmd/software_list/dyn-test-4
+    ...    message_pattern=.*successful.*
+    ${workflow_log}    Execute Command    cat /var/log/tedge/agent/workflow-software_list-dyn-test-4.log
+    Should Contain
+    ...    ${workflow_log}
+    ...    item="@version":"76e9afe834b4a7cadc9029670ba76745fcda73784f9e78c09f0c0416f7f58ad2"
+
+Recover Builtin Operation
+    ThinEdgeIO.File Should Exist    /etc/tedge/operations/software_list.toml
```
Comment (Bravo555) [inline tests/RobotFramework/tests/tedge_agent/workflows/dynamic_workflow_reloading.robot:88]: suggestion: the test fails if run by itself and not part of the test suite, because this file is created by the previous test case. Could we instead use Transfer to Device everywhere so that there are no dependencies between the test cases? I'd expect Transfer to Device to overwrite the file if it already exists, so it should be okay to use that?

File: crates/core/tedge_api/src/workflow/supervisor.rs [hunk 16]
```diff
@@ -229,6 +298,138 @@ impl WorkflowSupervisor {
     }
 }
 
+/// The set of in-use workflow versions for an operation
+///
+/// - The current version is the version that will be used for a new command instance.
+/// - The current version might be none. This is the case when the command has been deprecated.
+/// - When a new command instance is initialized, the current version is stored as being in use.
+/// - When all the commands using a given version are finalized, these copies are removed (TODO).
```
Comment (Bravo555) [inline crates/core/tedge_api/src/workflow/supervisor.rs:306]: TODO: does ae56d97652 implement removal mentioned here? If so, `TODO` can be removed

File: crates/core/tedge_agent/src/operation_workflows/persist.rs [hunk 17]
```diff
@@ -0,0 +1,478 @@
+use anyhow::Context;
+use camino::Utf8Path;
+use camino::Utf8PathBuf;
+use log::error;
```
Comment (Bravo555) [inline crates/core/tedge_agent/src/operation_workflows/persist.rs:4]:
```suggestion
use tracing::error;
```

File: crates/core/tedge_agent/src/operation_workflows/persist.rs
Comment (Bravo555) [inline crates/core/tedge_agent/src/operation_workflows/persist.rs:1]: thought: this module has quite a bit of functionality but doesn't have any unit tests - codecov reports 210 missed lines and 34.2% patch coverage (from other workflow-related tests)

**Review by Bravo555 — APPROVED** [review]
Some nits, but LGTM overall.

File: crates/core/tedge_api/src/workflow/supervisor.rs (hunk 16 above)
Comment (didier-wenzek) [reply crates/core/tedge_api/src/workflow/supervisor.rs]: fixed: https://github.com/thin-edge/thin-edge.io/pull/3180/commits/3b7d30eb8d3a38f0efad50b2a650816078a66b0f

File: crates/core/tedge_agent/src/operation_workflows/persist.rs (hunk 17 above)
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/operation_workflows/persist.rs]: Fixed https://github.com/thin-edge/thin-edge.io/pull/3180/commits/41f4208c93e8d51f8e43ee6dd95994d64472143d

File: crates/core/tedge_agent/src/operation_workflows/persist.rs
Comment (didier-wenzek) [reply crates/core/tedge_agent/src/operation_workflows/persist.rs:1]: I can only acknowledge that the *unit* test coverage is poor. However, this code is quit extensively tested by the system test suite added by this PR (Dynamic Workflow Reloading). I opted for system-tests instead of unit-tests because the features introduced by this module are heavily related to the file system and inotify as well as sequence of user actions (adding/updating/removing files while the agent is running/restarted. One place where unit tests can be improved is the `tedge_api::workflow::supervisor` module which provide the in-memory representation of the uploaded workflow definitions.

File: tests/RobotFramework/tests/tedge_agent/workflows/dynamic_workflow_reloading.robot (hunk 15 above)
Comment (didier-wenzek) [reply tests/RobotFramework/tests/tedge_agent/workflows/dynamic_workflow_reloading.robot:88]: I have a mix feeling here. On one side, you are correct, it would be handy to have independent tests. But, on the other side, this test suite represents well a scenario where a user creates and iterate updating a workflow file.

Concretely, replacing`File Should Exist` assertion by a `Transfer to Device` command would lead to a *different* test while running the suite vs the isolated test. Indeed, in the suite case, one checks that a user can update a workflow while, in the isolated case, one checks that the user can create a workflow (i.e. `Update User-Defined Operation` doing the same test as `Create User-Defined Operation`).
